File: samples/ocr_e2e/doc_img_orient_cls.cpp

/*
 * Copyright (C) 2026 Vastai-tech Company.
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
#include "common/doc_img_orient_cls.hpp"

#include <sstream>

#include "common/cmdline.hpp"
#include "common/model_profiler.hpp"

cmdline::parser ArgumentParser(int argc, char** argv) {
  cmdline::parser args;
  args.add<std::string>("model_prefix", 'm',
                        "model prefix of the model suite files", false,
                        "/opt/vastai/vaststreamx/data/models/"
                        "doc_ori_int8_mse_1-3-224-224/mod");
  args.add<std::string>("hw_config", '\0', "hw-config file of the model suite",
                        false);
  args.add<std::string>("vdsp_params", '\0', "vdsp preprocess parameter file",
                        false, "../data/configs/doc_ori_rgbplanar.json");
  args.add<uint32_t>("device_id", 'd', "device id to run", false, 0);
  args.add<std::string>("label_file", '\0', "label file", false, "[0, 180]");
  args.add<std::string>("input_file", '\0', "input image file", false,
                        "../data/images/ppocr-rot.jpg");
  args.add<std::string>("dataset_val_file", '\0', "dataset validation file",
                        false, "");
  args.add<std::string>("dataset_root", '\0', "input dataset root", false, "");
  args.parse_check(argc, argv);
  return args;
}

int get_angle(int index, std::vector<std::vector<int>>& labels) {
  for (auto& label : labels) {
    if (label[0] == index) {
      return label[1];
    }
  }
  std::cerr << "Error: Cann't find index: " << index << " in label file"
            << std::endl;
  return -10000;
}

int main(int argc, char** argv) {
  auto args = ArgumentParser(argc, argv);
  auto lines = vsx::LoadLabels(args.get<std::string>("label_file"));
  std::vector<std::vector<int>> labels;
  for (auto& line : lines) {
    int index, angle;
    std::istringstream iss(line);
    if (iss >> index >> angle)
      labels.push_back({index, angle});
    else {
      std::cerr << "Parsing label file Failed. line:" << line << std::endl;
      return -1;
    }
  }
  const int batch_size = 1;

  vsx::DocImgOrientClassifier model(args.get<std::string>("model_prefix"),
                                    args.get<std::string>("vdsp_params"),
                                    batch_size, args.get<uint32_t>("device_id"),
                                    args.get<std::string>("hw_config"));

  auto image_format = model.GetFusionOpIimageFormat();

  if (args.get<std::string>("dataset_val_file").empty()) {
    vsx::Image vsx_image;
    vsx::MakeVsxImage(args.get<std::string>("input_file"), vsx_image,
                      image_format);
    auto tensor = model.Process(vsx_image);
    int index;
    float score;
    model.PostProcess(tensor, index, score);
    std::cout << "Image angle: " << get_angle(index, labels)
              << ", confidence: " << score << std::endl;
  } else {
    std::vector<std::string> val_lines =
        vsx::ReadFileList(args.get<std::string>("dataset_val_file"));

    std::vector<std::string> filelist;
    std::vector<int> gt_indies;
    for (auto& line : val_lines) {
      std::istringstream iss(line);
      std::string file;
      int index;
      if (iss >> file >> index) {
        filelist.push_back(file);
        gt_indies.push_back(index);
      } else {
        std::cerr << "Parsing dataset val file Failed. line:" << line
                  << std::endl;
        return -1;
      }
    }
    auto dataset_root = args.get<std::string>("dataset_root");
    int correct_count = 0;
    for (size_t i = 0; i < filelist.size(); i++) {
      auto fullname = filelist[i];
      if (!dataset_root.empty()) fullname = dataset_root + "/" + filelist[i];
      std::cout << fullname << std::endl;
      vsx::Image vsx_image;
      vsx::MakeVsxImage(fullname, vsx_image, image_format);
      auto tensor = model.Process(vsx_image);
      int index;
      float score;
      model.PostProcess(tensor, index, score);
      if (index == gt_indies[i]) {
        correct_count++;
      }
      std::cout << "Predicted index: " << index
                << ", GT index: " << gt_indies[i]
                << ", correct_count: " << correct_count << std::endl;
    }

    std::cout << "Accuracy: " << correct_count / float(filelist.size())
              << std::endl;
  }

  return 0;
}

File: samples/ppocr/doc_img_orient_cls.py
#
# Copyright (C) 2026 Vastai-tech Company.
# All rights reserved.
#
# This source code is licensed under the MIT license found in the
# LICENSE file in the root directory of this source tree.
#

import os
import sys

current_file_path = os.path.dirname(os.path.abspath(__file__))
common_path = os.path.join(current_file_path, "../..")
sys.path.append(common_path)

from common.doc_img_orient_cls import DocImgOrientClassifier
import common.utils as utils

import cv2
import argparse


def argument_parser():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "-m",
        "--model_prefix",
        default="/opt/vastai/vaststreamx/data/models/doc_ori_int8_mse_1-3-224-224/mod",
        help="model prefix of the model suite files",
    )
    parser.add_argument(
        "--hw_config",
        default="",
        help="hw-config file of the model suite",
    )
    parser.add_argument(
        "--vdsp_params",
        default="./data/configs/doc_ori_bgr888.json",
        help="vdsp preprocess parameter file",
    )
    parser.add_argument(
        "-d",
        "--device_id",
        default=0,
        type=int,
        help="device id to run",
    )
    parser.add_argument(
        "--label_file",
        default="data/labels/ppocr_cls_labels.txt",
        help="label file",
    )
    parser.add_argument(
        "--input_file",
        default="data/images/img_rot180.jpg",
        help="input file",
    )
    parser.add_argument(
        "--dataset_val_file",
        default="",
        help="dataset validation file",
    )
    parser.add_argument(
        "--dataset_root",
        default="",
        help="input dataset root",
    )
    args = parser.parse_args()
    return args


if __name__ == "__main__":
    args = argument_parser()
    lines = utils.load_labels(args.label_file)
    labels={}
    for line in lines:
        label, angle = line.strip().split()
        labels[int(label)] = int(angle)
    
    batch_size = 1
    model = DocImgOrientClassifier(args.model_prefix, args.vdsp_params, batch_size, args.device_id)
    image_format = model.get_fusion_op_iimage_format()

    if args.dataset_val_file == "":
        cv_image = cv2.imread(args.input_file)
        assert cv_image is not None, f"Failed to read input file: {args.input_file}"
        vsx_image = utils.cv_bgr888_to_vsximage(cv_image, image_format, args.device_id)
        index, score = model.process(vsx_image)
        print(
            f"Image angle: {labels[index]}, confidence: {score:.4f}"
        )
    else:
        filelist = []
        gt={}
        with open(args.dataset_val_file, "rt") as f:
            lines = f.readlines()
            for line in lines:
                filename, label = line.strip().split()
                filelist.append(filename)
                gt[filename] = int(label)        
        correct_count = 0
    
        for file in filelist:
            fullname = os.path.join(args.dataset_root, file)
            print(fullname)
            cv_image = cv2.imread(fullname)
            assert cv_image is not None, f"Failed to read input file: {fullname}"
            vsx_image = utils.cv_bgr888_to_vsximage(
                cv_image, image_format, args.device_id
            )
            index, score = model.process(vsx_image)
            if index == gt.get(file, -1):
                correct_count += 1
            print(f"Image: {file}, Label: {gt.get(file, -1)}, Predicted: {index}, Correct: {correct_count}")
        print(f"Accuracy: {correct_count / len(filelist):.4f}")
